98_validate_binary_search_tree/step2.py

```diff
@@ -0,0 +1,38 @@
+from typing import Optional
+from math import inf
+from collections import deque
+
+class TreeNode:
+    def __init__(self, val=0, left=None, right=None):
+        self.val = val
+        self.left = left
+        self.right = right
+
+# step1と同様、根からDFSでチェックしていく
+class Solution:
+    def isValidBST(self, root: Optional[TreeNode]) -> bool:
+        def check_nodes_recursively(node, lower_bound, upper_bound):
+            if not node:
+                return True
+            if not lower_bound < node.val < upper_bound:
+                return False
+            if not check_nodes_recursively(node.left, lower_bound, node.val):
+                return False
+            return check_nodes_recursively(node.right, node.val, upper_bound)
```

Comment:
```suggestion
            if not check_nodes_recursively(node.left, lower_bound, node.val):
                return False
            if check_nodes_recursively(node.right, node.val, upper_bound)
                return False
            return True
```
と迷いました。基本的に
```
if P:
    return False
return True
```
は冗長かと思うのですが、今回のケースでは対称性があって予想しやすいコードになるような気もしており・・・

Reply:
私は、これは選択肢としてありだと思います。
逆に、return A and B も手ですね。